airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py

```diff
@@ -183,6 +187,40 @@ def open_file(self, file: RemoteFile, mode: FileReadMode, encoding: Optional[str
         # we can simply return the result here as it is a context manager itself that will release all resources
         return result
 
+    @override
+    def get_file(self, file: RemoteFile, local_directory: str, logger: logging.Logger):
+        file_size = self.file_size(file)
+        # I'm putting this check here so we can remove the safety wheels per connector when ready.
+        if file_size > self.FILE_SIZE_LIMIT:
+            message = "File size exceeds the 1 GB limit."
+            raise FileSizeLimitError(message=message, internal_message=message, failure_type=FailureType.config_error)
+
+        # Remove left slashes from source path format to make relative path for writing locally
+        file_relative_path = file.uri.lstrip("/")
+        local_file_path = path.join(local_directory, file_relative_path)
+
+        # Ensure the local directory exists
+        makedirs(path.dirname(local_file_path), exist_ok=True)
+
+        absolute_file_path = path.abspath(local_file_path)
+        logger.info(
+            f"Starting to download the file {file.uri} with size: {file_size / (1024 * 1024):,.2f} MB ({file_size / (1024 * 1024 * 1024):.2f} GB)"
+        )
+        # at some moment maybe we will require to play with the max_pool_connections and max_concurrency of s3 config
+        start_download_time = time.time()
+        self.s3_client.download_file(self.config.bucket, file.uri, local_file_path)
+        write_duration = time.time() - start_download_time
+        logger.info(f"Finished downloading the file {file.uri} and saved to {local_file_path} in {write_duration:,.2f} seconds.")
+        return {"file_url": absolute_file_path, "bytes": file_size, "file_relative_path": file_relative_path}
```

Comment:
Do we also share somewhere the original full URL?

Like `s3://my_bucket/path/to/file`?

Reply:
Do you mean for the record? We do it here in the CDK [here](https://github.com/airbytehq/airbyte/blob/497be1c2857615f4e3cab86e9ee09c451063eda2/airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py#L104).